ibm/resource_ibm_securitygroup_rule.go

```diff
@@ -0,0 +1,171 @@
+				Required: true,
+				ForceNew: true,
+			},
+		},
+	}
+}
+
+func findMatchingRule(d *schema.ResourceData,
+	service services.Network_SecurityGroup) (*datatypes.Network_SecurityGroup_Rule, error) {
+
+	direction := d.Get("direction").(string)
+	ethertype := d.Get("ether_type").(string)
+
+	portrangemin := d.Get("port_range_min").(int)
+	portrangemax := d.Get("port_range_max").(int)
+	if portrangemin < portrangemax {
+		return nil, fmt.Errorf("port range max should be greater or equal to min")
+	}
+
+	protocol := d.Get("protocol").(string)
+
+	securitygroupid := d.Get("security_group_id").(int)
+	log.Printf("[INFO] %s %d %d %s %d", ethertype, portrangemin, portrangemax, protocol, securitygroupid)
+
+	rules, err := service.Id(securitygroupid).GetRules()
+	if err != nil {
+		return nil, fmt.Errorf("Error fetching information for Security Group Rule: %s", err)
+	}
+
+	for _, rule := range rules {
+		if direction == *rule.Direction &&
+			ethertype == *rule.Ethertype &&
+			portrangemin == *rule.PortRangeMin &&
+			portrangemax == *rule.PortRangeMax &&
+			protocol == *rule.Protocol {
+			return &rule, nil
+		}
+	}
+
+	return nil, nil
+}
+
+func resourceIBMSecurityGroupRuleCreate(d *schema.ResourceData, meta interface{}) error {
+	sess := meta.(ClientSession).SoftLayerSession()
+	service := services.GetNetworkSecurityGroupService(sess)
+
+	matchingrule, err := findMatchingRule(d, service)
+	if err != nil {
+		return err
+	}
+
+	if matchingrule != nil {
+		log.Printf("[INFO] rule exists")
+		d.SetId(fmt.Sprintf("%d", *matchingrule.Id))
+		return nil
+	}
+
+	opts := []datatypes.Network_SecurityGroup_Rule{
+		{
+			Direction:    sl.String(d.Get("direction").(string)),
+			Ethertype:    sl.String(d.Get("ether_type").(string)),
+			PortRangeMin: sl.Int(d.Get("port_range_min").(int)),
+			PortRangeMax: sl.Int(d.Get("port_range_max").(int)),
+			Protocol:     sl.String(d.Get("protocol").(string)),
+		},
+	}
+	log.Println("[INFO] creating security group rule")
+	securitygroupid := d.Get("security_group_id").(int)
+	_, err = service.Id(securitygroupid).AddRules(opts)
+	if err != nil {
+		return fmt.Errorf("Error creating Security Group Rule: %s", err)
+	}
+
+	return resourceIBMSecurityGroupRuleRead(d, meta)
+}
+
+func resourceIBMSecurityGroupRuleRead(d *schema.ResourceData, meta interface{}) error {
+	sess := meta.(ClientSession).SoftLayerSession()
+	service := services.GetNetworkSecurityGroupService(sess)
+
+	matchingrule, err := findMatchingRule(d, service)
+	if err != nil {
+		return err
+	}
+
+	if matchingrule == nil {
+		return fmt.Errorf("Could not retrieve matching Security Group Rule")
+	}
+
+	d.SetId(strconv.Itoa(*matchingrule.Id))
+
+	d.Set("id", matchingrule.Id)
+	return nil
+}
+
+func resourceIBMSecurityGroupRuleDelete(d *schema.ResourceData, meta interface{}) error {
+	sess := meta.(ClientSession).SoftLayerSession()
+	service := services.GetNetworkSecurityGroupService(sess)
+
+	matchingrule, err := findMatchingRule(d, service)
```

Comment:
I don't think you need this. It can be considered as an optimization to see if the rule already exists before deleting.  ` _, err = service.Id(*matchingrule.SecurityGroupId).RemoveRules(ids_to_delete)` should suffice.